### ISSUE TITLE
SSL_new() fails for QUIC and OPENSSL_THREADS_NONE

### ISSUE THREAD

**Issue by Futaura** [openssl/openssl#22982]
When `OPENSSL_THREADS` and `OPENSSL_THREADS_NONE` are both defined, `SSL_new()` fails for QUIC connections because of the following code:

https://github.com/openssl/openssl/blob/986c48c4eb26861f25bc68ea252d8f2aad592735/ssl/quic/quic_impl.c#L387-L392

Should the code really fail at that point if `OPENSSL_THREADS_NONE` is defined? In such a case, removing the those lines of code allows a QUIC connection to be successfully established.

For some context, in the OpenSSL port for AmiSSL, we have been using the existing thread/locking functions, as per `crypto/threads*.c` for many years, but for 3.2.0 we currently are building with thread pools disabled and no threaded implementation for QUIC.

**Comment by hlandau** [comment]
Can you explain how it is possible to have `OPENSSL_THREADS` and `OPENSSL_THREADS_NONE`? This is supposed to be set automatically by thread_arch.h, not manually.

It does not make sense to have a build which has `OPENSSL_THREADS` but no `ossl_crypto_mutex_new`. Perhaps you want to define `OPENSSL_NO_THREAD_POOL` instead (via `./Configure no-thread-pool`)?

**Comment by Futaura** [comment]
Yes, sorry, what I meant was that `OPENSSL_THREADS` is defined and because there is currently no OS support for the new `crypto/thread/arch` in our build, this causes `thread_arch.h` to define `OPENSSL_THREADS_NONE`. Configure is already run with `no-thread-pool` and [our configuration](https://github.com/jens-maus/amissl/blob/master/openssl/Configurations/15-amissl.conf) defines a `thread_scheme` which causes `OPENSSL_THREADS` to be defined.

I can of course patch this in our ported code, but it just seems strange that `OPENSSL_THREADS_NONE` is currently a bit pointless, as rather than being a no-op (like in `crypto/threads_none.c`), the code in `crypto/thread/arch/thread_none.c` is not a no-op as `ossl_crypto_mutex_new` can only fail (on the other hand `CRYPTO_THREAD_lock_new` returns success even in `threads_none.c`). Lack of consistency, basically.

**Comment by nhorman** [comment]
This still doesn't make any sense to me.  OPENSSL_THREADS_NONE and OPENSSL_THREADS should be mutually exlusive.  Are you suggesting that in your build both somehow become defined?

Looking at thread_arch.h:

```
# if defined(OPENSSL_THREADS) && defined(OPENSSL_SYS_UNIX)
#  define OPENSSL_THREADS_POSIX
# elif defined(OPENSSL_THREADS) && defined(OPENSSL_SYS_VMS)
#  define OPENSSL_THREADS_POSIX
# elif defined(OPENSSL_THREADS) && defined(OPENSSL_SYS_WINDOWS) && \
    defined(_WIN32_WINNT)
#  if _WIN32_WINNT >= 0x0600
#   define OPENSSL_THREADS_WINNT
#  elif _WIN32_WINNT >= 0x0501
#   define OPENSSL_THREADS_WINNT
#   define OPENSSL_THREADS_WINNT_LEGACY
#  else
#   define OPENSSL_THREADS_NONE
#  endif
# else
#  define OPENSSL_THREADS_NONE
# endif
```
OPENSSL_THREADS_NONE should only be defined iff OPENSSL_THREADS is not defined and all of OPENSSL_SYS_UNIX/OPENSSL_SYS_VMS/OPENSSL_SYS_WINDOWS is not defined.

I can imagine that happening if you are building for a new platform that is none of UNIX/VMS/WINDOWS.  If that is the case, it would seem to me that part of your platform support should be to properly define one of those, or create your own.

Looking at your initial post, it seems you are trying to build AMISSL, which some quick googling suggests is an openssl port to AmigaOS.  is that the case?

If so, first, thats really neat!  That said, I'm afraid thats an unsupported platform, and likely not something we will be able to help you with, unless you want to submit a pull request to add support, which I would encourage you to do.

**Comment by t8m** [comment]
IMO, the above code doesn't preclude both OPENSSL_THREADS and OPENSSL_THREADS_NONE being defined at once. However we should immediately error out in such situation. You should build with `no-threads` if there is no thread support for your platform.

**Comment by t8m** [comment]
What you describe is `no-thread-pool` (OPENSSL_NO_THREAD_POOL). If you want your openssl build to never create threads but keep support for multithreaded applications you just need to pass `no-thread-pool` to `Configure` instead of `no-threads`.

**Comment by t8m** [comment]
Please note we are looking for community help implementing this issue. Please comment if you are interested in implementing this feature or if you have reasons why this feature is important enough so we should consider implementing it with higher priority than other requested features. The issue will be closed after 2 months otherwise.

**Comment by t8m** [comment]
This issue was closed because it was marked with the help wanted label and there was no interest in implementing this issue or any other activity leading to implementing or re-prioritizing it for more than 2 months.
